api/envoy/config/listener/v3/listener.proto

```diff
@@ -90,6 +90,14 @@ message Listener {
   // Configuration for envoy internal listener. All the future internal listener features should be added here.
   // [#not-implemented-hide:]
   message InternalListenerConfig {
+    // The max contiguous chained internal listener to internal cluster is allowed. 0 means upstream
+    // connection cannot use internal connection. The goal is to avoid a bad config that containing
+    // loop of the internal listener and internal cluster. Such a loop will OOM the Envoy in a second.
+    // Per listener connection limit doesn't work well here because it takes 1 dispatcher cycle to reach
+    // the per listener connection limit but may take N cycles to remove the long chain of internal
+    // connections.
+    // [#not-implemented-hide:]
+    google.protobuf.UInt32Value hop_to_live = 1;
```

Comment:
How do you plan on implementing this?

Reply:
The hop_to_live(HTL) value can be attached to the stream info. The HTL can be propagated from downstream connection and decreased when creating upstream internal connection. 
The upstream internal connection is always created by pair e.g. the C1 and C2: 
C0 ListenerA <->Cluster  C1 <-> C2 ListenerB. 
C0 is the downstream connection of C1
C2 is the pair connection of C2

Suppose listenerA is the frist internal listener, we set C0.htl = hop_to_live
C1.htl := C0.htl - 1
C2.htl := C0.htl - 1
If the C2's htl < 0, the C1 and C2 connection will be closed immediately.

This is for L4 internal connection as stream... For http stream it's viable as well although we probably need to introduce other fake internal header names, something like retry-timeout.

Reply:
Is the plan to automatically propagate the C0 stream info across to C2? Will there be a "lastHopStreamInfo()" method or something?

Reply:
Yes, streamInfo should have hop_value() returning optional<uint>. "lastHopStreamInfo()" is true IFF hop_value() is optional(0).
Only if the streaminfo returns nullopt, the internal listener should set the new value in the listener config.

transport socket option of C1 can be added by downstream connection C0. 
When C1 is created, the internal connection manager creates the connection socket for C2.
When the internal listener promotes the connection socket from to the connection, the listener checks the HTL.

It adds tax for each connection but this is easier than validate the xDS configs to detect loop.